src/test_application/flight.h

```diff
@@ -0,0 +1,40 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2020 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "OSPMagnum.h"
+
+#include <thread>
+
+/**
+ * Start flight scene. This will create an ActiveArea in the universe and start
+ * a Magnum application with an ActiveScene. All the necessary systems will be
+ * registered to this scenes for in-universe flight, such as SysAreaAssociate.
+ *
+ * @param magnumApp [out] Magnum application created
+ * @param OSPApp [in,out] OSP universe and resources to run the application on
+ * @param args [in] Arguments to pass to Magnum
+ */
```

Comment:
In this situation you should probably be returning the unique pointer by value, and not using an "out" parameter.

Reply:
I'm not sure how well this fits with best practices, but this function is used to start a new thread. It uses the unique_ptr to pass the application it creates back to the main thread.

Reply:
My point was just that you can change the function from

void foo(std::unique_ptr<>& pOut);

to

std::unique_ptr<> foo(void);

Returning a std::unique_ptr by value is basically free, and literally free if you write the function to take advantage of return value optimization.

Accepting a reference to an std::unique_ptr, and then assigning a value to that passed in reference, can't be optimized out in any meaningful way.

Same outcome, different ways of doing it.

Plus, if you return the std::unique_ptr, you can mark the function as [[nodiscard]] and get a nice warning if the variable isn't used after being received.

Reply:
I should put this in the doxygen block; this function blocks execution until the window is closed. It blocks at `pMagnumApp->exec();` The function will not return anything for quite a while.